File: PinetreeShop/PinetreeShop.Domain.Baskets/BasketAggregate.cs
using PinetreeShop.CQRS.Infrastructure;
using PinetreeShop.Domain.Baskets.Events;
using PinetreeShop.Domain.Baskets.Commands;
using PinetreeShop.Domain.Baskets.Exceptions;
using System.Collections.Generic;
using System.Linq;
using PinetreeShop.Domain.Shared.Types;
using PinetreeShop.Domain.Shared.Exceptions;

namespace PinetreeShop.Domain.Baskets
{
    public class BasketAggregate : AggregateBase
    {
        public enum BasketState { Pending, Cancelled, CheckedOut };
        public BasketState State { get; private set; }
        public List<OrderLine> OrderLines { get; private set; }

        #region Contructor

        private BasketAggregate(CreateBasket cmd) : this()
        {
            RaiseEvent(new BasketCreated(cmd.AggregateId));
        }

        public BasketAggregate()
        {
            OrderLines = new List<OrderLine>();
            State = BasketState.Pending;

            RegisterEventHandler<BasketCreated>(Apply);
            RegisterEventHandler<BasketItemAdded>(Apply);
            RegisterEventHandler<BasketItemRemoved>(Apply);
            RegisterEventHandler<BasketCancelled>(Apply);
            RegisterEventHandler<BasketCheckedOut>(Apply);
        }

        #endregion

        #region Event handlers

        private void Apply(BasketCreated evt)
        {
            AggregateId = evt.AggregateId;
            State = BasketState.Pending;
        }

        private void Apply(BasketItemAdded evt)
        {
            AddProductToOrderLines(evt);
        }

        private void Apply(BasketItemRemoved evt)
        {
            RemoveProductFromOrderLines(evt);
        }

        private void Apply(BasketCancelled evt)
        {
            State = BasketState.Cancelled;
        }

        private void Apply(BasketCheckedOut evt)
        {
            State = BasketState.CheckedOut;
        }

        #endregion

        #region Command handlers

        internal static BasketAggregate Create(CreateBasket cmd)
        {
            return new BasketAggregate(cmd);
        }

        internal void AddItemToBasket(AddItemToBasket cmd)
        {
            if (State != BasketState.Pending)
                throw new InvalidStateException(AggregateId, $"Cannot add item. Basket is {State}");

            RaiseEvent(new BasketItemAdded(cmd.AggregateId, cmd.ProductId, cmd.ProductName, cmd.Price, cmd.Quantity));
        }

        internal void RemoveItemFromBasket(RemoveItemFromBasket cmd)
        {
            if (State != BasketState.Pending) throw new InvalidStateException(AggregateId, $"Cannot remove item. Basket is {State}");

            var basketId = cmd.AggregateId;
            var productId = cmd.ProductId;
            var quantity = cmd.Quantity;

            var orderLine = OrderLines.SingleOrDefault(ol => ol.ProductId == productId);
            if (orderLine != null)
            {
                if (orderLine.Quantity < quantity) quantity = orderLine.Quantity;
                RaiseEvent(new BasketItemRemoved(basketId, productId, quantity));
            }
        }

        internal void Cancel(CancelBasket cmd)
        {
            if (State == BasketState.Cancelled) return;

            if (State != BasketState.Pending)
                throw new InvalidStateException(cmd.AggregateId, $"Cannot cancel, basket is {State}");

            RaiseEvent(new BasketCancelled(cmd.AggregateId));
        }

        internal void CheckOut(CheckOutBasket cmd)
        {
            if (cmd.ShippingAddress == null)
                throw new DomainException(cmd.AggregateId, $"Cannot check out, address is null");

            if (State == BasketState.CheckedOut || !OrderLines.Any()) return;

            if (State != BasketState.Pending)
                throw new InvalidStateException(cmd.AggregateId, $"Cannot check out, basket is {State}");

            RaiseEvent(new BasketCheckedOut(cmd.AggregateId, OrderLines, cmd.ShippingAddress));
        }

        #endregion

        #region Helpers

        private void AddProductToOrderLines(BasketItemAdded evt)
        {
            var productId = evt.ProductId;
            var productName = evt.ProductName;
            var price = evt.Price;
            var quantity = evt.Quantity;
            var orderLine = OrderLines.SingleOrDefault(ol => ol.ProductId == productId);
            if (orderLine == null)
            {
                orderLine = new OrderLine
                {
                    ProductId = productId,
                    ProductName = productName,
                    Price = price,
                    Quantity = quantity
                };
                OrderLines.Add(orderLine);
            }
            else
            {
                orderLine.Quantity += quantity;
            }
        }

        private void RemoveProductFromOrderLines(BasketItemRemoved evt)
        {
            var productId = evt.ProductId;
            var quantity = evt.Quantity;
            var orderLine = OrderLines.Single(ol => ol.ProductId == productId);
            orderLine.Quantity = orderLine.Quantity - quantity;

            if (orderLine.Quantity == 0)
            {
                OrderLines.Remove(orderLine);
            }
        }

        #endregion
    }
}

File: PinetreeShop/PinetreeShop.Domain.Baskets/BasketAggregate.fs
module PinetreeShop.Domain.Baskets.BasketAggregate

open PinetreeCQRS.Infrastructure.Types
open PinetreeCQRS.Infrastructure.Commands
open PinetreeCQRS.Infrastructure.Events
open System
open Chessie.ErrorHandling
open PinetreeCQRS.Infrastructure.Validation

type BasketError =
    | ValidationError of string
    interface IError
    override e.ToString() = sprintf "%A" e
    
type BasketState =
    | NotCreated
    | Pending
    | Cancelled
    | CheckedOut

type Command = 
    | Create
    | AddItem of BasketItem 
    | RemoveItem of ProductId * int
    | Cancel
    | CheckOut of ShippingAddress
    interface ICommand

type Event = 
    | BasketCreated
    | BasketItemAdded of BasketItem
    | BasketItemRemoved of ProductId * int
    | BasketCancelled
    | BasketCheckedOut of ShippingAddress * BasketItem list
    interface IEvent

type State = 
    { BasketState : BasketState
      Items : Map<ProductId, BasketItem> }
    static member Zero = 
        { BasketState = NotCreated
          Items = Map.empty }

module private Handlers = 

    let addItem currentItems (item:BasketItem) : Map<ProductId, BasketItem>= 
        let currentItem = Map.tryFind item.ProductId currentItems 
        let newItem = 
            match currentItem with
            | None -> item
            | Some i -> { i with Quantity = i.Quantity + item.Quantity }
        Map.add newItem.ProductId newItem currentItems              

    let removeItem currentItems productId quantity : Map<ProductId, BasketItem> =     
        let currentItem = Map.tryFind productId currentItems 
        match currentItem with
        | None -> currentItems
        | Some i ->
            if (i.Quantity - quantity) <= 0 then Map.remove productId currentItems
            else Map.add productId { i with Quantity = i.Quantity - quantity } currentItems


    let applyEvent state event = 
        match event with
        | BasketCreated -> { state with BasketState = Pending }
        | BasketItemAdded item -> { state with Items = addItem state.Items item }
        | BasketItemRemoved (productId, qty) -> { state with Items = removeItem state.Items productId qty }
        | BasketCancelled -> { state with BasketState = Cancelled }
        | BasketCheckedOut _ -> { state with BasketState = CheckedOut }


    module private Validate =
        let inCase predicate error value = 
            let result = predicate value
            match result with
            | false -> ok value
            | true -> Bad [ ValidationError error :> IError ]

        module private Helpers =
            let notEmptyShippingAddress (ShippingAddress sa) = 
                inCase (fun sa -> String.IsNullOrWhiteSpace(sa)) "Shipping Address cannot be empty" sa
            let isBasketState s os = 
                inCase (fun s -> s.BasketState <> os) (sprintf "Wrong Basket state %A" s.BasketState) s
            let notBasketState s os = 
                inCase (fun s -> s.BasketState = os) (sprintf "Wrong Basket state %A" s.BasketState) s
            let hasItems ol = inCase (fun ol -> ol = Map.empty) "No items" ol
            let canCreate s = isBasketState s NotCreated
            let created s = notBasketState s NotCreated
        
        let canCreate s = Helpers.canCreate s
        let canCheckOut s sa = Helpers.isBasketState s Pending <* Helpers.hasItems s.Items <* Helpers.notEmptyShippingAddress sa
        let canCancel s = Helpers.isBasketState s Pending 
        let canAddItem s = Helpers.isBasketState s Pending
        let canRemoveItem s = Helpers.isBasketState s Pending


    let executeCommand (state:State) command = 
        match command with
        | Create -> Validate.canCreate state <?> [ BasketCreated ]
        | AddItem item -> Validate.canAddItem state <?> [ BasketItemAdded item ]
        | RemoveItem (productId, qty) -> 
            let removeItems =
                let item = Map.tryFind productId state.Items
                match item with
                | Some i ->
                    if(i.Quantity <= qty) then [BasketItemRemoved(productId, i.Quantity)]
                    else [ BasketItemRemoved(productId, qty)]
                | None -> []
            Validate.canRemoveItem state <?> removeItems
        | Cancel -> Validate.canCancel state <?> [ BasketCancelled ]
        | CheckOut sa -> 
            let items = Map.toList state.Items |>  List.map (fun (k, v) -> v)
            Validate.canCheckOut state sa <?> [ BasketCheckedOut(sa, items) ]


let makeBasketCommandHandler = 
    makeCommandHandler { Zero = State.Zero
                         ApplyEvent = Handlers.applyEvent
                         ExecuteCommand = Handlers.executeCommand } 

let loadBasket events = Seq.fold Handlers.applyEvent State.Zero events